### PR TITLE
Bug 1725935: Split handling of degraded across multiple conditions

### PR DESCRIPTION
**Pull request by enj**
This change is meant as a starting point for more nuanced handling of the degraded condition.  We will likely need to add more conditions as the sync loop is broken apart.

Signed-off-by: Monis Khan <mkhan@redhat.com>

xref: #141

### CODE REVIEW

**Comment by enj** [comment]
/hold

I want to see how CI fares + do some manual testing.

**Comment by enj** [comment]
e2e-aws looks promising:

```
$ oc get authentication.operator cluster -o yaml
```

```yaml
apiVersion: operator.openshift.io/v1
kind: Authentication
metadata:
  annotations:
    release.openshift.io/create-only: "true"
  creationTimestamp: "2019-07-01T19:31:35Z"
  generation: 1
  name: cluster
  resourceVersion: "16172"
  selfLink: /apis/operator.openshift.io/v1/authentications/cluster
  uid: d6efe738-9c36-11e9-9cb3-1264a6a33b12
spec:
  managementState: Managed
status:
  conditions:
  - lastTransitionTime: "2019-07-01T19:37:40Z"
    status: "False"
    type: ResourceSyncControllerDegraded
  - lastTransitionTime: "2019-07-01T19:38:57Z"
    status: "False"
    type: RouteStatusDegraded
  - lastTransitionTime: "2019-07-01T19:42:48Z"
    status: "False"
    type: OperatorSyncDegraded
  - lastTransitionTime: "2019-07-01T19:39:01Z"
    status: "False"
    type: IdentityProviderConfigDegraded
  - lastTransitionTime: "2019-07-01T19:42:48Z"
    status: "False"
    type: RouteHealthDegraded
  - lastTransitionTime: "2019-07-01T19:42:40Z"
    status: "False"
    type: WellKnownDegraded
  - lastTransitionTime: "2019-07-01T19:50:41Z"
    status: "False"
    type: Progressing
  - lastTransitionTime: "2019-07-01T19:50:41Z"
    reason: AsExpected
    status: "True"
    type: Available
  - lastTransitionTime: "2019-07-01T19:50:41Z"
    status: "False"
    type: OAuthClientsDegraded
  generations:
  - group: apps
    hash: ""
    lastGeneration: 1
    name: oauth-openshift
    namespace: openshift-authentication
    resource: deployments
  observedGeneration: 1
  readyReplicas: 2
```

```
$ oc get co authentication -o yaml
```

```yaml
apiVersion: config.openshift.io/v1
kind: ClusterOperator
metadata:
  creationTimestamp: "2019-07-01T19:37:40Z"
  generation: 1
  name: authentication
  resourceVersion: "16175"
  selfLink: /apis/config.openshift.io/v1/clusteroperators/authentication
  uid: aff2a278-9c37-11e9-a596-1282413a18fc
spec: {}
status:
  conditions:
  - lastTransitionTime: "2019-07-01T19:42:40Z"
    reason: AsExpected
    status: "False"
    type: Degraded
  - lastTransitionTime: "2019-07-01T19:50:41Z"
    reason: AsExpected
    status: "False"
    type: Progressing
  - lastTransitionTime: "2019-07-01T19:50:41Z"
    reason: AsExpected
    status: "True"
    type: Available
  - lastTransitionTime: "2019-07-01T19:37:40Z"
    reason: NoData
    status: Unknown
    type: Upgradeable
  extension: null
  relatedObjects:
  - group: operator.openshift.io
    name: cluster
    resource: authentications
  - group: config.openshift.io
    name: cluster
    resource: authentications
  - group: config.openshift.io
    name: cluster
    resource: infrastructures
  - group: config.openshift.io
    name: cluster
    resource: oauths
  - group: ""
    name: openshift-config
    resource: namespaces
  - group: ""
    name: openshift-config-managed
    resource: namespaces
  - group: ""
    name: openshift-authentication
    resource: namespaces
  - group: ""
    name: openshift-authentication-operator
    resource: namespaces
  versions:
  - name: operator
    version: 0.0.1-2019-07-01-191604
  - name: oauth-openshift
    version: 0.0.1-2019-07-01-191604_openshift
```

**Comment by enj** [comment]
/hold cancel

Per CI results.

File: pkg/operator2/operator.go [hunk 1]
```diff
@@ -460,6 +457,7 @@ func (c *authOperator) handleVersion(
 	}
 
 	wellknownReady, wellknownMsg, err := c.checkWellknownEndpointsReady(authConfig, route)
+	handleDegraded(operatorConfig, "WellKnown", err)
```
[the remark this answers is not in the document]
Comment (sallyom) [reply pkg/operator2/operator.go]: +1

File: pkg/operator2/operator.go [hunk 2]
```diff
@@ -252,13 +252,8 @@ func (c *authOperator) Sync(obj metav1.Object) error {
 
 	operatorConfigCopy := operatorConfig.DeepCopy()
 
-	// clear degraded status
-	setDegradedFalse(operatorConfigCopy)
```
Comment (sallyom) [inline pkg/operator2/operator.go]: glad to see this go, didn't seem right here

**Comment by enj** [comment]
Improved handling of degraded to avoid multiple conditions unless actually needed:

```yaml
apiVersion: config.openshift.io/v1
kind: ClusterOperator
metadata:
  creationTimestamp: "2019-07-01T00:59:23Z"
  generation: 1
  name: authentication
  resourceVersion: "557784"
  selfLink: /apis/config.openshift.io/v1/clusteroperators/authentication
  uid: 775ff7db-9b9b-11e9-a62c-0a0097dfad48
spec: {}
status:
  conditions:
  - lastTransitionTime: "2019-07-02T14:45:50Z"
    message: 'RouteStatusDegraded: route is not available at canonical host oauth-openshift.bapps.somal630.devcluster.openshift.com:
      [{Host:oauth-openshift.apps.somal630.devcluster.openshift.com RouterName:default
      Conditions:[{Type:Admitted Status:True Reason: Message: LastTransitionTime:2019-07-01
      01:04:54 +0000 UTC}] WildcardPolicy:None RouterCanonicalHostname:apps.somal630.devcluster.openshift.com}]'
    reason: RouteStatusDegradedError
    status: "True"
    type: Degraded
  - lastTransitionTime: "2019-07-02T14:38:26Z"
    reason: AsExpected
    status: "False"
    type: Progressing
  - lastTransitionTime: "2019-07-01T01:11:03Z"
    reason: AsExpected
    status: "True"
    type: Available
  - lastTransitionTime: "2019-07-01T00:59:23Z"
    reason: NoData
    status: Unknown
    type: Upgradeable
  extension: null
  relatedObjects:
  - group: operator.openshift.io
    name: cluster
    resource: authentications
  - group: config.openshift.io
    name: cluster
    resource: authentications
  - group: config.openshift.io
    name: cluster
    resource: infrastructures
  - group: config.openshift.io
    name: cluster
    resource: oauths
  - group: ""
    name: openshift-config
    resource: namespaces
  - group: ""
    name: openshift-config-managed
    resource: namespaces
  - group: ""
    name: openshift-authentication
    resource: namespaces
  - group: ""
    name: openshift-authentication-operator
    resource: namespaces
  versions:
  - name: operator
    version: 4.2.0-0.ci-2019-06-30-044256
  - name: oauth-openshift
    version: 4.2.0-0.ci-2019-06-30-044256_openshift
```

```yaml
apiVersion: operator.openshift.io/v1
kind: Authentication
metadata:
  annotations:
    release.openshift.io/create-only: "true"
  creationTimestamp: "2019-07-01T00:53:46Z"
  generation: 1
  name: cluster
  resourceVersion: "557783"
  selfLink: /apis/operator.openshift.io/v1/authentications/cluster
  uid: aeaf360c-9b9a-11e9-92df-024d74c0f9f4
spec:
  managementState: Managed
status:
  conditions:
  - lastTransitionTime: "2019-07-01T00:59:23Z"
    status: "False"
    type: ResourceSyncControllerDegraded
  - lastTransitionTime: "2019-07-01T01:11:03Z"
    status: "False"
    type: Degraded
  - lastTransitionTime: "2019-07-02T14:38:26Z"
    status: "False"
    type: Progressing
  - lastTransitionTime: "2019-07-01T01:11:03Z"
    reason: AsExpected
    status: "True"
    type: Available
  - lastTransitionTime: "2019-07-02T14:44:50Z"
    message: 'route is not available at canonical host oauth-openshift.bapps.somal630.devcluster.openshift.com:
      [{Host:oauth-openshift.apps.somal630.devcluster.openshift.com RouterName:default
      Conditions:[{Type:Admitted Status:True Reason: Message: LastTransitionTime:2019-07-01
      01:04:54 +0000 UTC}] WildcardPolicy:None RouterCanonicalHostname:apps.somal630.devcluster.openshift.com}]'
    reason: Error
    status: "True"
    type: RouteStatusDegraded
  - lastTransitionTime: "2019-07-02T14:37:42Z"
    status: "False"
    type: IdentityProviderConfigDegraded
  - lastTransitionTime: "2019-07-02T14:38:26Z"
    status: "False"
    type: RouteHealthDegraded
  - lastTransitionTime: "2019-07-02T14:37:42Z"
    status: "False"
    type: WellKnownEndpointDegraded
  - lastTransitionTime: "2019-07-02T14:37:42Z"
    status: "False"
    type: OAuthClientsDegraded
  - lastTransitionTime: "2019-07-02T15:45:32Z"
    status: "False"
    type: OperatorSyncDegraded
  generations:
  - group: apps
    hash: ""
    lastGeneration: 2
    name: oauth-openshift
    namespace: openshift-authentication
    resource: deployments
  observedGeneration: 1
  readyReplicas: 2
```

**Comment by enj** [comment]
/retest

Terraform

**Comment by smarterclayton** [comment]
This is super novel.  Who did API review on this?

**Comment by smarterclayton** [comment]
I would have expected you to report Degraded, a reason (if there was just one) and a message that summarized multiple errors (see cluster version operator).  This approach is super novel, and I'm not comfortable with this without a lot more thought.  Crawfords concerns on pruning degraded is a big one, but each condition is an API.  You are trying really hard to return conditions, but none of the tools in the system will look at the sub conditions, so they're mostly wasted.

I'll review more tomorrow but I need some justification to allow this sort of complexity.

**Comment by sttts** [comment]
@smarterclayton @enj isn't this the same the other control plane operators do (for "ages"), encoded in library-go? I.e. having separate `*Degraded` condition on the main operator CRD and one aggregated one on the `ClusterOperator`?

**Comment by deads2k** [comment]
> This is super novel. Who did API review on this?

@smarterclayton Me.  This is not clusteroperator.config.openshift.io, these are conditions for authentication.*operator*.openshift.io.  The object that customers do not generally interact with.  This behavior is the same one we use to provide distinct information about which particular piece of the kube-apiserver operator is degraded.  Multiple conditions on a low-level operator resource are summarized into a single condition on the clusteroperator.config.openshift.io.

**Comment by smarterclayton** [comment]
Looking at this, you're using etcd as a synchronization primitive for internal state between internal loops.  I understand the chain of reasoning that led to this, I just question whether it's worth the cost.

This is acceptable for operator.openshift.io resources but not for config.openshift.io resources, so as long as it stays in that group I'm fine with it.  I'll remove the hold from the other PR.

**Pull request merged**